File: leetcode/container-with-most-water.java
//https://leetcode.com/problems/container-with-most-water/

//Slightly Optimized Syntax
class Solution {
    public int maxArea(int[] height) {
        int maxArea = 0;
        int left = 0;
        int right = height.length-1;
        
        while(left < right){
            // Equivalent to Math.max(maxArea, (right-left) * Math.min(height[left], height[right]))
            maxArea = maxArea < (height[left] < height[right] ? height[left] : height[right])*(right-left) ? (height[left] < height[right] ? height[left] : height[right])*(right-left) : maxArea;
            if(height[left]<height[right]) left++;
            else right--;
        }
        
        return maxArea;
    }
}

/* Clean to read solution
class Solution {
    public int maxArea(int[] height) {
        int maxArea = 0;
        int left = 0;
        int right = height.length-1;
        int currArea = 0;
        
        while(left<right){
            if(height[left] < height[right]){
                currArea = (right-left) * height[left];
                left++;
            }else{
                currArea = (right-left) * height[right];
                right--;
            }
            maxArea = Math.max(currArea, maxArea);
        }
        
        return maxArea;
    }
}
*/

File: leetcode/container_with_most_water.py
# https://leetcode.com/problems/container-with-most-water/description/


class Solution:
    def maxArea(self, height: list[int]) -> int:
        left: int = 0
        right: int = len(height) - 1
        most: int = 0

        while left < right:
            min_height = min(height[left], height[right])
            most = max(most, (right - left) * min_height)
            if height[left] <= height[right]:
                left += 1
            else:
                right -= 1

        return most
